docs/modularity.rst

```diff
@@ -0,0 +1,194 @@
+at both the process and pipeline level. 
+
+Below we have four methods including example to illustrate the different approaches.
+
+- External scripts
+- Templating
+- Nextflow in Nextflow with GitHub
+- Nextflow in Nextflow with local files
+
+.. _modularity-script:
+
+External Scripts
+=======
+One simple way to modularise processes is to have the `script` section of the 
+process as an external script.
+
+In Nextflow, the script section can be composed of any programming language.
+For example, in BASH we you can be either write out the commands in the script section
+::
+
+  process sayHello {
+
+      """
+      echo 'Hello world!' > myFile.txt
+      cat myFile.txt
+      """
+
+  }
+  
+Or encapsuslate them in an external script. In this example we can create an external shell
+script named `helloWorld.sh` within the `bin` directory:
+::
+    #!/usr/bin/bash
+    echo 'Hello world!' > myFile.txt
+    cat myFile.txt
+  
+And then call this file in the process: 
+::
+    process sayHello {
+
+      """
+      helloWorld.sh
+      """
+
+  }
+  
+
+As a more real-world example, consider the `sleuth` process in the Kallisto-NF pipe. 
+Here the process calls an external R script [here](https://github.com/cbcrg/kallisto-nf/blob/master/bin/sleuth.R) with the arguments of the script (`$exp_file`) originating 
+from Nextflow channels:
+::
+    process sleuth {
+        input:
+        file 'kallisto/*' from kallisto_out_dirs.collect()     
+        file exp_file
+
+        output: 
+        file 'sleuth_object.so'
+        file 'gene_table_results.txt'
+
+        script:
+        """
+        sleuth.R kallisto ${exp_file}
+        """
+    }
+
+
+Templates
+=======
+
+Templates provide an alternative to external scripts and expose the input/output channel values to the template without the need to declare them again. This is very useful for when similar but different tasks can defined by the same process. 
+
+Consider an example below, where we wish to quantifiy transcripts with one (or more) of three possible pseudomapping tools. We can create a single pseudomapping process and have different templates for the specific commands of each tool. 
+::
+    process mapping_template {
+        input:
+        val(mapper) from ('kallisto', 'sailfish', 'salmon')
+        set val(sampleID), file(reads) from read_files
+        file index_dir from indexes.collect()
+
+        output: 
+        file "${mapper}_${sampleID}" into mapped_reads
+
+        script:
+        template mapping_${mapper}.sh
+    }
+    
+
+Inside the `templates` directory we could have three template files:
+
+- mapping_kallisto.sh
+- mapping_sailfish.sh
+- mapping_salmon.sh
+
+and the template file `templates/mapping_kallisto.nf` could contain:
+::
+    #!/usr/bin/bash
+    mkdir kallisto_${sampleID}
+    kallisto quant -i ${index_dir}/kallisto.index -o ${mapper}_${sampleID} ${reads}
+    
+whilst the template file `templates/mapping_salmon.nf` could contain:
```

Comment:
whilst the template file `templates/mapping_salmon.nf` could contain:

whilst the template file `templates/mapping_salmon.sh` could contain: